File: api/moderate.py
from http.server import BaseHTTPRequestHandler
import json
import os

# Configuration
ADMIN_SECRET = os.getenv('ADMIN_SECRET', 'your_secure_admin_secret_here')

# Try to import requests for Upstash REST API
HAS_REDIS = False
try:
    import requests
    REDIS_URL = os.getenv('KV_REST_API_URL')
    REDIS_TOKEN = os.getenv('KV_REST_API_TOKEN')
    
    if REDIS_URL and REDIS_TOKEN:
        HAS_REDIS = True
        print(f"Using Upstash Redis at {REDIS_URL}")
    else:
        print("Redis credentials not found in environment")
except ImportError:
    print("Warning: requests not available, using in-memory storage")

# In-memory storage fallback
_memory_storage = {}

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path.startswith('/api/comments/moderate'):
            # Check admin auth
            auth_header = self.headers.get('Authorization', '')
            if not auth_header.startswith('Bearer '):
                self.send_response(401)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Unauthorized'}).encode())
                return
            
            provided_secret = auth_header[7:]  # Remove 'Bearer ' prefix
            if provided_secret != ADMIN_SECRET:
                self.send_response(401)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Unauthorized'}).encode())
                return
            
            # Load all comments
            comments = load_comments()
            
            # Get all unapproved comments from all metals
            pending_comments = []
            for metal, metal_comments in comments.items():
                for comment in metal_comments:
                    if not comment.get('approved', False):
                        pending_comments.append(comment)
            
            # Sort by timestamp (newest first)
            pending_comments.sort(key=lambda x: x.get('timestamp', ''), reverse=True)
            
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps(pending_comments).encode())
    
    def do_POST(self):
        if self.path.startswith('/api/comments/moderate'):
            # Check admin auth
            auth_header = self.headers.get('Authorization', '')
            if not auth_header.startswith('Bearer '):
                self.send_response(401)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Unauthorized'}).encode())
                return
            
            provided_secret = auth_header[7:]  # Remove 'Bearer ' prefix
            if provided_secret != ADMIN_SECRET:
                self.send_response(401)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Unauthorized'}).encode())
                return
            
            # Read request body
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            
            try:
                mod_data = json.loads(post_data.decode('utf-8'))
            except json.JSONDecodeError:
                self.send_response(400)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Invalid JSON'}).encode())
                return
            
            comment_id = mod_data.get('commentId')
            metal_name = mod_data.get('metalName')
            action = mod_data.get('action')  # 'approve' or 'reject'
            
            if not all([comment_id, metal_name, action]):
                self.send_response(400)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Missing required fields'}).encode())
                return
            
            if action not in ['approve', 'reject']:
                self.send_response(400)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Invalid action'}).encode())
                return
            
            # Load comments
            comments = load_comments()
            
            if metal_name not in comments:
                self.send_response(404)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Metal not found'}).encode())
                return
            
            # Find and modify the comment
            comment_found = False
            for i, comment in enumerate(comments[metal_name]):
                if str(comment.get('id')) == str(comment_id):
                    if action == 'approve':
                        comments[metal_name][i]['approved'] = True
                    else:  # reject
                        comments[metal_name].pop(i)  # Remove rejected comments
                    comment_found = True
                    break
            
            if not comment_found:
                self.send_response(404)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Comment not found'}).encode())
                return
            
            # Save updated comments
            save_comments(comments)
            
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({
                'success': True,
                'message': f'Comment {action}d successfully'
            }).encode())
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
        self.end_headers()

def load_comments():
    """Load comments from Upstash Redis or memory"""
    global _memory_storage
    
    try:
        if HAS_REDIS:
            import requests
            url = f"{os.getenv('KV_REST_API_URL')}/get/comments"
            headers = {
                'Authorization': f"Bearer {os.getenv('KV_REST_API_TOKEN')}"
            }
            
            response = requests.get(url, headers=headers)
            if response.status_code == 200:
                data = response.json()
                result = data.get('result')
                if result:
                    return json.loads(result)
            return {}
        else:
            # Use in-memory storage for development
            return _memory_storage
    except Exception as e:
        print(f"Error loading comments: {e}")
        return {}

def save_comments(comments):
    """Save comments to Upstash Redis or memory"""
    global _memory_storage
    
    try:
        if HAS_REDIS:
            import requests
            url = f"{os.getenv('KV_REST_API_URL')}/set/comments"
            headers = {
                'Authorization': f"Bearer {os.getenv('KV_REST_API_TOKEN')}",
                'Content-Type': 'application/json'
            }
            
            # Send the JSON string directly as the body to Upstash
            response = requests.post(url, 
                                   data=json.dumps(comments),
                                   headers=headers)
            if response.status_code != 200:
                print(f"Failed to save comments: {response.text}")
        else:
            # Use in-memory storage for development
            _memory_storage = comments
    except Exception as e:
        print(f"Error saving comments: {e}")

File: api/comments/moderate.js
import { kv } from '@vercel/kv';

// Admin authentication from environment variables
const ADMIN_SECRET = process.env.ADMIN_SECRET;

export default async function handler(req, res) {
  // Enable CORS
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, DELETE, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  // Check authentication
  const authHeader = req.headers.authorization;
  if (!ADMIN_SECRET || !authHeader || authHeader !== `Bearer ${ADMIN_SECRET}`) {
    return res.status(401).json({ error: 'Unauthorized - Admin secret not configured or invalid' });
  }

  try {
    if (req.method === 'GET') {
      // Get all pending comments for moderation
      const allKeys = [];
      let cursor = '0';
      
      // Scan for all comment keys
      do {
        const result = await kv.scan(cursor, { match: 'comments:*' });
        cursor = result[0];
        allKeys.push(...result[1]);
      } while (cursor !== '0');

      const pendingComments = [];
      
      for (const key of allKeys) {
        const comments = await kv.get(key) || [];
        const metalName = key.replace('comments:', '');
        
        comments.forEach(comment => {
          if (!comment.approved) {
            pendingComments.push({
              ...comment,
              metalName,
              key
            });
          }
        });
      }

      // Sort by timestamp, newest first
      pendingComments.sort((a, b) => new Date(b.timestamp) - new Date(a.timestamp));

      res.status(200).json(pendingComments);
    }
    else if (req.method === 'POST') {
      // Approve or reject a comment
      const { commentId, action, metalName } = req.body;
      
      if (!commentId || !action || !metalName) {
        return res.status(400).json({ error: 'Missing required fields' });
      }

      const key = `comments:${metalName}`;
      const comments = await kv.get(key) || [];
      
      if (action === 'approve') {
        // Approve the comment
        const updated = comments.map(comment => 
          comment.id === commentId 
            ? { ...comment, approved: true, approvedAt: new Date().toISOString() }
            : comment
        );
        await kv.set(key, updated);
        
        res.status(200).json({ success: true, message: 'Comment approved' });
      }
      else if (action === 'reject') {
        // Remove the comment
        const filtered = comments.filter(comment => comment.id !== commentId);
        await kv.set(key, filtered);
        
        res.status(200).json({ success: true, message: 'Comment rejected and removed' });
      }
      else {
        res.status(400).json({ error: 'Invalid action' });
      }
    }
    else {
      res.status(405).json({ error: 'Method not allowed' });
    }
  } catch (error) {
    console.error('Moderation API Error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
}